### PR TITLE
allow failed discovery on initial quota controller start

### LINKED ISSUE #65005: kube-controller-manager fails to start if API discovery for extended API servers fails
/area controller-manager
/area extensibility
/kind bug
/sig apimachinery

**What happened**:
The kube-controller-manager is unable to start when it cannot discover the APIs of API server extensions. Precisely, the `APIService` object indicated availability although the extended API server was not running anymore. Restarting the kube-controller-manager failed for the mentioned reason.

**What you expected to happen**:
The `APIService` object should not indicate availability if the respective API server is not running. Not sure if the kube-controller-manager could ignore APIs it cannot discover, what does it actually need to do with them? Actually, the only way to make the kube-controller-manager start again is to temporarily de-register the extended API, start it, and register it again.

**How to reproduce it (as minimally and precisely as possible)**:
Deploy a Kubernetes cluster, register an aggregated API server, wait for the `APIService` status to indicate availability, then delete the aggregated API server pod and restart the kube-controller-manager. It will 

**Environment**:
- Kubernetes version (use `kubectl version`): 1.10.4

### PR DESCRIPTION
**Pull request by deads2k**
Fixes #65005

Aggregated API servers now correctly provide 503s on discovery endpoints for groups that cannot be reached.  This means that the kube-controller-manager process is now sensitive to discovery failures in the quota controller.  This change allows discovery failures in the initial quota replenishment controller resource discovery.

@liggitt suspects similar races exist to those he found GC last release, but this pull doesn't make that better or worse.

@kubernetes/sig-api-machinery-bugs 

```release-note
kube-controller-manager can now start the quota controller when discovery results can only be partially determined.
```

### CODE REVIEW

File: pkg/controller/resourcequota/resource_quota_controller.go [hunk 1]
```diff
@@ -161,10 +161,11 @@ func NewResourceQuotaController(options *ResourceQuotaControllerOptions) (*Resou
 
 		rq.quotaMonitor = qm
 
-		// do initial quota monitor setup
+		// do initial quota monitor setup.  If we have a discovery failure here, it's ok. We'll discover more resources when
+		// a later sync happens.
 		resources, err := GetQuotableResources(options.DiscoveryFunc)
 		if err != nil {
```
Comment (liggitt) [inline pkg/controller/resourcequota/resource_quota_controller.go]: should this only tolerate `IsGroupDiscoveryFailedError` errors?

File: pkg/controller/resourcequota/resource_quota_controller.go [hunk 2]
```diff
@@ -421,7 +422,9 @@ func (rq *ResourceQuotaController) Sync(discoveryFunc NamespacedResourcesFunc, p
 	// Something has changed, so track the new state and perform a sync.
 	oldResources := make(map[schema.GroupVersionResource]struct{})
 	wait.Until(func() {
-		// Get the current resource list from discovery.
+		// Get the current resource list from discovery.  A failure here will (should?) prevent updates to the sync list.
```
Comment (liggitt) [inline pkg/controller/resourcequota/resource_quota_controller.go]: Is tolerating partial discovery on start but not on resync what we want? I expected them to match. `ErrGroupDiscoveryFailed` errors expose which group/versions failed, if we want to use that info

Comment (liggitt) [reply pkg/controller/resourcequota/resource_quota_controller.go]: even without inspecting the details of the error, we could do something like this as a simple start. letting the quota controller continue running without using available resource info from discovery doesn't seem great.

```go
// Get the current resource list from discovery.
newResources, err := GetQuotableResources(discoveryFunc)
if err != nil {
	utilruntime.HandleError(err)

	if discovery.IsGroupDiscoveryFailedError(err) && len(newResources) > 0 {
		// In partial discovery cases, don't remove any existing informers, just add new ones
		for k, v := range oldResources {
			newResources[k] = v
		}
	} else {
		// short circuit in non-discovery error cases or if discovery returned zero resources
		return
	}
}
```

**Comment by deads2k** [comment]
@liggitt comments addressed.

**Comment by nilebox** [comment]
or @deads2k: cherry-pick to `release-1.11` and `release-1.10` is needed, see above.

**Comment by liggitt** [comment]
added to https://github.com/kubernetes/kubernetes/pull/67155 and https://github.com/kubernetes/kubernetes/pull/67154

**Pull request merged**